File: src/android/ClarityPlugin.java
package com.microsoft.clarity.cordova;

import android.app.Activity;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.RequiresApi;
import com.microsoft.clarity.Clarity;
import com.microsoft.clarity.ClarityConfig;
import com.microsoft.clarity.models.ApplicationFramework;
import com.microsoft.clarity.models.LogLevel;
import java.util.ArrayList;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.FutureTask;
import java.util.concurrent.RunnableFuture;
import org.apache.cordova.*;
import org.json.JSONArray;
import org.json.JSONException;


@RequiresApi(api = Build.VERSION_CODES.Q)
public class ClarityPlugin extends CordovaPlugin {
    @Override
    public boolean execute(String action, JSONArray data, CallbackContext callbackContext) {
        try {
            switch (action) {
                case "initialize":
                    if (initialize(data)) {
                        callbackContext.success("Clarity initialized.");
                    } else {
                        callbackContext.error("Failed to initialize Clarity, domain may not be allowed. Please check logs for more details!");
                    }
                    return true;
                case "pause":
                    Clarity.pause();
                    if(Clarity.isPaused()) {
                        callbackContext.success("Clarity paused.");
                    } else {
                        callbackContext.error("Failed to pause Clarity, please check logs for more details!");
                    }
                    return true;
                case "resume":
                    Clarity.resume();
                    if(!Clarity.isPaused()) {
                        callbackContext.success("Resume succeeded.");
                    } else {
                        callbackContext.error("Failed to resume Clarity, please check logs for more details!");
                    }
                    return true;
                case "isPaused":
                    // CallbackContext.success doesn't accept Boolean.
                    callbackContext.success(Clarity.isPaused() ? 1 : 0);
                    return true;
                case "setCustomUserId":
                    String customUserId = data.get(0).equals(null) ? null : data.getString(0);

                    if (Clarity.setCustomUserId(customUserId)) {
                        callbackContext.success("Setting custom user id succeeded.");
                    } else {
                        callbackContext.error("Setting custom user id failed, please check logs for more details!");
                    }
                    return true;
                case "setCustomSessionId":
                    String customSessionId = data.get(0).equals(null) ? null : data.getString(0);

                    if (Clarity.setCustomSessionId(customSessionId)) {
                        callbackContext.success("Setting custom session id succeeded.");
                    } else {
                        callbackContext.error("Setting custom session id failed, please check logs for more details!");
                    }
                    return true;
                case "setCurrentScreenName":
                    String currentScreenName = data.get(0).equals(null) ? null : data.getString(0);

                    if (Clarity.setCurrentScreenName(currentScreenName)) {
                        callbackContext.success("Setting custom screen name succeeded.");
                    } else {
                        callbackContext.error("Setting custom screen name failed, please check logs for more details!");
                    }
                    return true;
                case "setCustomTag":
                    String key = data.get(0).equals(null) ? null : data.getString(0);
                    String value = data.get(1).equals(null) ? null : data.getString(1);

                    if (Clarity.setCustomTag(key, value)) {
                        callbackContext.success("Setting custom tag succeeded.");
                    } else {
                        callbackContext.error("Setting custom tag failed, please check logs for more details!");
                    }
                    return true;
                case "getCurrentSessionId":
                    callbackContext.success(Clarity.getCurrentSessionId());
                    return true;
                case "getCurrentSessionUrl":
                    callbackContext.success(Clarity.getCurrentSessionUrl());
                    return true;
            }

        } catch (Exception e) {
            callbackContext.error(e.getMessage());
        }

        return false;
    }

    private boolean initialize(JSONArray data) throws JSONException {
        Activity activity = this.cordova.getActivity();

        String projectId = data.getString(0);
        LogLevel logLevel = LogLevel.valueOf(data.getString(2));
        Boolean isIonic = data.getBoolean(7);
        Boolean enableWebViewCapture = true; // If false, nothing is going to be captured.
        ApplicationFramework applicationFramework = !isIonic ? ApplicationFramework.Cordova : ApplicationFramework.Ionic;


        ClarityConfig config = new ClarityConfig(
            projectId
        );
        config.setLogLevel(logLevel);
        config.setApplicationFramework(applicationFramework);

        Callable<Boolean> callable = () -> {
            return Clarity.initialize(activity, config);
        };

        RunnableFuture<Boolean> task = new FutureTask<>(callable);

        // Run on the main thread as recommended by the native Clarity SDK.
        new Handler(Looper.getMainLooper()).post(task);
        try {
            return task.get(); // this will block until Runnable completes
        } catch (InterruptedException | ExecutionException e) {
            return false;
        }
    }

    private ArrayList<String> jsonArrayToList(JSONArray arr) throws JSONException {
        ArrayList<String> ret = new ArrayList<String>();

        for (int i = 0; i < arr.length(); i++) {
            ret.add(arr.getString(i));
        }

        return ret;
    }
}

File: src/ios/ClarityPlugin.swift
import Foundation
import Clarity

@objc(ClarityPlugin)
class ClarityPlugin: CDVPlugin {
    
    @objc(initialize:)
    func initialize(command: CDVInvokedUrlCommand) {
        guard let projectId = command.arguments[0] as? String,
              let logLevel = command.arguments[2] as? Int else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid parameters")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        let clarityLogLevel: LogLevel = LogLevel(rawValue: logLevel) ?? .none
        let framework: ApplicationFramework = .cordova
        
        let clarityConfig = ClarityConfig(
            projectId: projectId,
            logLevel: clarityLogLevel,
            applicationFramework: framework
        )
        
        if ClaritySDK.initialize(config: clarityConfig) {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Clarity initialized.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Failed to initialize Clarity, please check logs for more details!")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    @objc(pause:)
    func pause(command: CDVInvokedUrlCommand) {
        ClaritySDK.pause()
        if ClaritySDK.isPaused() {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Clarity paused.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Failed to pause Clarity, please check logs for more details!")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    @objc(resume:)
    func resume(command: CDVInvokedUrlCommand) {
        ClaritySDK.resume()
        if !ClaritySDK.isPaused() {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Resume succeeded.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Failed to resume Clarity, please check logs for more details!")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    @objc(isPaused:)
    func isPaused(command: CDVInvokedUrlCommand) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: ClaritySDK.isPaused() ? 1 : 0)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(setCustomUserId:)
    func setCustomUserId(command: CDVInvokedUrlCommand) {
        guard let customUserId = command.arguments[0] as? String else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid custom user id")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        if ClaritySDK.setCustomUserId(customUserId) {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Setting custom user id succeeded.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Setting custom user id failed, please check logs for more details!")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    @objc(setCustomSessionId:)
    func setCustomSessionId(command: CDVInvokedUrlCommand) {
        guard let customSessionId = command.arguments[0] as? String else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid custom session id")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        if ClaritySDK.setCustomSessionId(customSessionId) {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Setting custom session id succeeded.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Setting custom session id failed, please check logs for more details!")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    @objc(setCurrentScreenName:)
    func setCurrentScreenName(command: CDVInvokedUrlCommand) {
        guard let currentScreenName = command.arguments[0] as? String else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid screen name")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        if ClaritySDK.setCurrentScreenName(currentScreenName) {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Setting custom screen name succeeded.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Setting custom screen name failed, please check logs for more details!")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    @objc(setCustomTag:)
    func setCustomTag(command: CDVInvokedUrlCommand) {
        guard let key = command.arguments[0] as? String,
              let value = command.arguments[1] as? String else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Invalid key or value")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        if ClaritySDK.setCustomTag(key: key, value: value) {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Setting custom tag succeeded.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Setting custom tag failed, please check logs for more details!")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    @objc(getCurrentSessionId:)
    func getCurrentSessionId(command: CDVInvokedUrlCommand) {
        let sessionId = ClaritySDK.getCurrentSessionId()
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: sessionId)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getCurrentSessionUrl:)
    func getCurrentSessionUrl(command: CDVInvokedUrlCommand) {
        let sessionUrl = ClaritySDK.getCurrentSessionUrl()
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: sessionUrl)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
}
